8255243: Reinforce escape barrier interactions with ZGC conc stack processing

fisk opened the pull request:
The escape barrier reallocates scalarized objects potentially deep into the stack of a remote thread. Each allocation can safepoint, causing referenced frames to be invalid. Some sprinklings were added that deal with that, but I believe it was subsequently broken with the integration of the new vector API, that has its own new deoptimization code that did not know about this. Not surprisingly, the integration of the new vector API had no idea about this subtlety, and allocates an object, and then reads an object deep from the stack of a remote thread (using an escape barrier). I suppose the issue is that all these 3 things were integrated at almost the same time. The problematic code sequence is in VectorSupport::allocate_vector() in vectorSupport.cpp, which is called from Deoptimization::realloc_objects(). It first allocates an oop (possibly safepointing), and then reads a vector oop from the stack. This is usually fine, but not through the escape barrier, with concurrent stack scanning. While I have not seen any crashes yet, I can see from code inspection, that there is no way that this works correctly.

In order to make this less fragile for future changes, we should really have a RAII object that keeps the target thread's stack of the escape barrier, stable and processed, across safepoints. This patch fixes that. Then it becomes much easier to reason about its correctness, compared to hoping the various hooks are applied after each safepoint.

With this new robustness fix, the thread running the escape barrier, keeps the target thread stack processed, straight through safepoints on the requesting thread, making it easy and intuitive to understand why this works correctly. The RAII object basically just has to cover the code block that pokes at the remote stack and goes in and out of safepoints, arbitrarily. Arguably, this escape barrier doesn't need to be blazingly fast, and can afford keeping stacks sane through its operation.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x32 | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ❌ (1/9 failed) | ✔️ (9/9 passed) |

**Failed test task**
- [Windows x64 (hs/tier1 compiler)](https://github.com/fisk/jdk/runs/1315264950)

### Issue
 * [JDK-8255243](https://bugs.openjdk.java.net/browse/JDK-8255243): Reinforce escape barrier interactions with ZGC conc stack processing


### Reviewers
 * [Richard Reingruber](https://openjdk.java.net/census#rrich) (@reinrich - Committer)
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)


### Contributors
 * Richard Reingruber `<rrich@openjdk.org>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/832/head:pull/832`
`$ git checkout pull/832`

fisk commented:
@reinrich Since you wrote the escape barrier, I thought I'd ping you in case you are interested. This makes the mechanism more robust and reliable, w.r.t. concurrent stack processing. No issue observed, but looks wrong with the new vector API deoptimization code that was integrated recently. This makes it more future-proof and easy to reason about.

fisk commented:
Forgot to mention, but I tested this from tier1-5, to sanity check that the new solution doesn't introduce any new issues.

reinrich commented:
@fisk thanks for pinging. I remember reading about a new vector API a while ago and noticed that it potentially interferes with escape barriers but then I lost track. I'll look at this.

dcubed-ojdk commented:
Thanks for adding the tier testing info. Sounds like there is no need for new
tests since this is "just" a RAII object to reinforce existing code interactions,
but please confirm that...

fisk commented:
> Thanks for adding the tier testing info. Sounds like there is no need for new
> 
> tests since this is "just" a RAII object to reinforce existing code interactions,
> 
> but please confirm that...

Exactly; confirmed.

reinrich commented on src/hotspot/share/runtime/keepStackGCProcessed.cpp:
```diff
@@ -0,0 +1,60 @@
+    return;
+  }
+  StackWatermark* their_watermark = StackWatermarkSet::get(jt, StackWatermarkKind::gc);
+  our_watermark->link_watermark(their_watermark);
```
Assert our_watermark->_linked_watermark == NULL to avoid unintentional nesting?

reinrich reviewed (changes requested):
I'm really glad you caught that one! And I like the abstraction provided by KeepStackGCProcessedMark.

There is one execution path you missed coming from `VM_GetOrSetLocal::deoptimize_objects(javaVFrame* jvf)`. This code should probably be moved into EscapeBarrier. `EscapeBarrier::deoptimize_objects(int depth)` could be changed to be more generic `EscapeBarrier::deoptimize_objects(int from_depth, int to_depth)`. VM_GetOrSetLocal::doit_prologue() could call eb.deoptimize_objects(_depth, _depth) then. That would be better but maybe not yet really good...

Thanks again, Richard.

reinrich commented:
Hi Erik, the last commit in https://github.com/reinrich/jdk/commits/pr-832-with-better-encapsulation would be the refactoring I would like to do. It removes the code not compliant with concurrent thread stack processing from VM_GetOrSetLocal::doit_prologue(). Instead EscapeBarrier::deoptimize_objects(int d1, int d2) is called. You added already a KeepStackGCProcessedMark to that method and I changed it to accept a range [d1, d2] of frames do the object deoptimization for.

I'm not sure how to handle this from a process point of view. Can the refactoring be done within this change? Should a new item or subtask be created for it. I'd be glad if you could give an advice on that.

Thanks, Richard.

fisk commented:
> Hi Erik, the last commit in https://github.com/reinrich/jdk/commits/pr-832-with-better-encapsulation would be the refactoring I would like to do. It removes the code not compliant with concurrent thread stack processing from VM_GetOrSetLocal::doit_prologue(). Instead EscapeBarrier::deoptimize_objects(int d1, int d2) is called. You added already a KeepStackGCProcessedMark to that method and I changed it to accept a range [d1, d2] of frames do the object deoptimization for.
> 
> I'm not sure how to handle this from a process point of view. Can the refactoring be done within this change? Should a new item or subtask be created for it. I'd be glad if you could give an advice on that.
> 
> Thanks, Richard.

If you are okay with it, I can add your refactorings into this change, and add you as a co-author of the change. Sounds good?

Thanks,
/Erik

reinrich commented:
It does sound good indeed to me if you don't mind doing that. Thanks!
I have run the tests dedicated to EscapeBarriers with ZGC enabled and also the DeoptimizeObjectsALot stress testing. I will run some more serviceability tests and my teams CI testing until tomorrow.

fisk commented:
Thanks @reinrich. I uploaded your patch to this PR, and will add you as contributor. Also addressed your review comments. Hope your testing went fine.

reinrich reviewed (approved):
Thanks for making EscapeBarriers more robust with regard to concurrent thread stack processing.
The change looks good to me.

fisk commented:
/contributor add @reinrich

sspitsyn reviewed (approved):
Hi Erik and Richard,
Changes in the serviceability files looks fine.
Thanks,
Serguei

fisk commented:
> Hi Erik and Richard,
> 
> Changes in the serviceability files looks fine.
> 
> Thanks,
> 
> Serguei
> 
> 

Thanks for the review Serguei!

fisk commented:
/integrate
